### PR TITLE
#1317 Add support for ignoring unmapped source properties

### LINKED ISSUE #1317: Add support for ignoring source properties follow up on #610
With the integration of #1282 we have support for reporting on unmapped source properties. Additionally we need to add support for ignoring certain source properties.

We need to do the following:

1. Make the `target()` parameter of `@Mapping` optional, so that we can simply use `@Mapping(source="foo", ignore=true)`
1. `target()` should only be allowed to be empty when `unmappedSourcePolicy != IGNORE`, `ignore = true` and `source` has some value, otherwise it should not be allowed.
1. Ignoring `source` should only be allowed hen `unmappedSourcePolicy != IGNORE`

### PR DESCRIPTION
**Pull request by filiphr**
This property has only effect on the unmapped source properties report

Fixes #1317

### CODE REVIEW

File: core-common/src/main/java/org/mapstruct/BeanMapping.java [hunk 1]
```diff
@@ -83,4 +83,18 @@
      * @since 1.3
      */
     boolean ignoreByDefault() default false;
+
+    /**
+     * The unmapped source properties that need to be ignored. This should be used when {@link ReportingPolicy#WARN}
```
Comment (sjaakd) [inline core-common/src/main/java/org/mapstruct/BeanMapping.java]: should == could?

File: core-common/src/main/java/org/mapstruct/BeanMapping.java [hunk 2]
```diff
@@ -83,4 +83,18 @@
...
+     * or {@link ReportingPolicy#ERROR} is used for {@link Mapper#unmappedSourcePolicy()} or
+     * {@link MapperConfig#unmappedSourcePolicy()}. When properties are set they won't be part of the unmapped source
+     * properties report. This property has no effect on choosing whether a source property needs to be mapped.
+     * <p>
+     * NB: This does not support ignoring nested source properties
```
Comment (sjaakd) [inline core-common/src/main/java/org/mapstruct/BeanMapping.java]: NB = NOTE?

File: core-common/src/main/java/org/mapstruct/BeanMapping.java (hunk 1 above)
Comment (sjaakd) [inline core-common/src/main/java/org/mapstruct/BeanMapping.java]: Unmapped source properties to be ignored.

File: core-common/src/main/java/org/mapstruct/BeanMapping.java [hunk 3]
```diff
@@ -83,4 +83,18 @@
      * @since 1.3
      */
     boolean ignoreByDefault() default false;
+
+    /**
+     * The unmapped source properties that need to be ignored. This should be used when {@link ReportingPolicy#WARN}
+     * or {@link ReportingPolicy#ERROR} is used for {@link Mapper#unmappedSourcePolicy()} or
+     * {@link MapperConfig#unmappedSourcePolicy()}. When properties are set they won't be part of the unmapped source
```
Comment (sjaakd) [inline core-common/src/main/java/org/mapstruct/BeanMapping.java]: Listed properties will be ignored when composing the unmapped source properties report.

File: core-common/src/main/java/org/mapstruct/BeanMapping.java [hunk 4]
```diff
@@ -83,4 +83,18 @@
...
+     * properties report. This property has no effect on choosing whether a source property needs to be mapped.
```
Comment (sjaakd) [inline core-common/src/main/java/org/mapstruct/BeanMapping.java]: I don't understand.. `This properties has.. mapped.`

**Comment by filiphr** [comment]
Thanks for the review @sjaakd I've pushed a modified wording.

**Comment by filiphr** [comment]
Thanks for the review @sjaakd

**Pull request merged**